modules/tinymce/src/themes/silver/main/ts/ui/core/ChoiceControls.ts

```diff
@@ -0,0 +1,145 @@
+    const lastApi = Singleton.destroyable();
+
+    const callback = () => {
+      const current = spec.getCurrent(editor);
+      const apiOpt = current.map(spec.hash).bind((key) => Obj.get(apis, key));
+      apiOpt.fold(
+        // If we don't have a menu item for the current state, make sure we're not highlighting anything
+        lastApi.clear,
+        // If we do have a menu item for the current state, highlight it
+        (api) => {
+          lastApi.set({
+            destroy: () => api.setActive(false)
+          });
+          api.setActive(true);
+        }
+      );
+    };
+
+    editor.on('NodeChange', callback);
+
+    return Arr.map(
+      options,
+      (value, i) => ({
+        type: 'togglemenuitem',
+        text: spec.display(value),
+        onSetup: (api) => {
+          apis[spec.hash(value)] = api;
+
+          if (i + 1 === options.length) {
+            // run the callback once on startup (on the last option so that we know the apis map has been set up)
+            callback();
+          }
+
+          return () => {
+            // only clean up global things once
+            if (i === 0) {
+              editor.off('NodeChange', callback);
+              lastApi.clear();
+            }
+          };
+        },
+        onAction: () => spec.setCurrent(editor, value)
+      })
+    );
+  };
+
+  editor.ui.registry.addMenuButton(spec.name, {
+    tooltip: spec.text,
+    icon: spec.icon,
+    fetch: (callback) => callback(getMenuItems())
+  });
+
+  editor.ui.registry.addNestedMenuItem(spec.name, {
+    type: 'nestedmenuitem',
+    text: spec.text,
+    getSubmenuItems: getMenuItems
+  });
+};
+
+const lineHeightSpec: ControlSpec<string> = {
+  name: 'lineheight',
+  text: 'Line height',
+  icon: 'line-height',
+
+  getOptions: Settings.getLineHeightFormats,
+  hash: (input) => Dimension.normalise(input, [ 'fixed', 'relative', 'empty' ]).getOr(input),
+  display: Fun.identity,
+
+  getCurrent: (editor) => Optional.from(editor.queryCommandValue('LineHeight')),
+  setCurrent: (editor, value) => editor.execCommand('LineHeight', false, value)
+};
+
+const languageSpec: ControlSpec<ContentLanguage> = {
+  name: 'language',
+  text: 'Language',
+  icon: 'translate',
+
+  getOptions: Settings.getContentLanguages,
+  hash: (input) => Type.isUndefined(input.customCode) ? input.code : `${input.code}/${input.customCode}`,
+  display: (input) => input.title,
+
+  getCurrent: (editor) => {
+    const node = SugarElement.fromDom(editor.selection.getNode());
+    return TransformFind.closest(node, (n) =>
+      Optional.some(n)
+        .filter(SugarNode.isElement)
+        .bind((ele) => {
+          const codeOpt = Attribute.getOpt(ele, 'lang');
+          const customCode = Attribute.getOpt(ele, 'data-mce-lang').getOrUndefined();
+          return codeOpt.map((code): ContentLanguage => ({
+            code,
+            customCode,
+            title: ''
+          }));
```

Comment:
```suggestion
          return codeOpt.map((code): ContentLanguage => ({
            code,
            customCode: Attribute.getOpt(ele, 'data-mce-lang').getOrUndefined(),
            title: ''
          }));
```
This way we're not creating an Optional and looking up the attribute when we don't need to be.

Reply:
If you prefer, we could also do something like this:
```suggestion
          return codeOpt.map((code): ContentLanguage => {
            const customCode = Attribute.getOpt(ele, 'data-mce-lang').getOrUndefined();
            return { code, customCode, title: '' };
          });
```